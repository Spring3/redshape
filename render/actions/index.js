import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
