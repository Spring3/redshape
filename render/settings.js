export const availableOptions = {
  // Used in OptionsBlock.jsx and settings.reducer.js
  // uiStyle, idleBehavior, progressSlider, timerCheckpoint
  uiStyle: [
    { value: 'default', label: 'Default' },
    { value: 'colors', label: 'Default with colors' },
    { value: 'enhanced', label: 'Enhanced' },
  ],
  idleBehavior: [
    { value: 'none', label: 'Do nothing' },
    { value: '1m', label: 'Pause if idle for 1 minute' },
    { value: '2m', label: 'Pause if idle for 2 minutes' },
    { value: '5m', label: 'Pause if idle for 5 minutes' },
    { value: '10m', label: 'Pause if idle for 10 minutes' },
    { value: '15m', label: 'Pause if idle for 15 minutes' },
    { value: '20m', label: 'Pause if idle for 20 minutes' },
    { value: '30m', label: 'Pause if idle for 30 minutes' },
    { value: '45m', label: 'Pause if idle for 45 minutes' },
    { value: '1h', label: 'Pause if idle for 1 hour' },
  ],
  progressSlider: [
    { value: '1%', label: 'Slider with 1% steps' },
    { value: '2%', label: 'Slider with 2% steps' },
    { value: '5%', label: 'Slider with 5% steps' },
    { value: '10%', label: 'Slider with 10% steps' },
  ],
  timerCheckpoint: [
    { value: 'none', label: 'Do not save state periodically' },
    { value: '1m', label: 'Save state every minute' },
    { value: '2m', label: 'Save state every 2 minutes' },
    { value: '3m', label: 'Save state every 3 minutes' },
    { value: '5m', label: 'Save state every 5 minutes' },
    { value: '10m', label: 'Save state every 10 minutes' },
    { value: '15m', label: 'Save state every 15 minutes' },
    { value: '30m', label: 'Save state every 30 minutes' },
  ],
  // Used in ColumnHeadersSelect.jsx and settings.reducer.js
  // issueHeaders
  issueHeaders: [
    {
      label: 'Id', isFixed: true, value: 'id', format: 'id'
    },
    { label: 'Project', value: 'project.name' },
    { label: 'Tracker', value: 'tracker.name', format: 'tracker' },
    { label: 'Status', value: 'status.name', format: 'status' },
    { label: 'Subject', isFixed: true, value: 'subject' },
    { label: 'Priority', value: 'priority.name', format: 'priority' },
    { label: 'Estimation', value: 'estimated_hours', format: 'hours' },
    { label: 'Total Estimation', value: 'total_estimated_hours', format: 'hours' },
    { label: 'Spent', value: 'spent_hours', format: 'hours' },
    { label: 'Total Spent', value: 'total_spent_hours', format: 'hours' },
    { label: 'Progress', value: 'done_ratio', format: 'progress' },
    { label: 'Tags', value: 'tags', format: 'tags' },
    { label: 'Due Date', value: 'due_date', format: 'date' },
    { label: 'Attachments', value: 'attachments', format: 'count' },
    { label: 'Assigned', value: 'assigned_to.name' },
    { label: 'Author', value: 'author.name' },
    { label: 'Updated', value: 'updated_on', format: 'date' },
    { label: 'Created', value: 'created_on', format: 'date' },
  ]
};
