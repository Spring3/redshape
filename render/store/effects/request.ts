import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
