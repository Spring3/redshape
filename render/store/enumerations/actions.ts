import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
