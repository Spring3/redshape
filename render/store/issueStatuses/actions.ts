import { IAction, Context } from 'overmind';
import { IssueStatus, PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

const getAll: IAction<void, Promise<PaginatedActionResponse<IssueStatus>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'issue_statuses.json'
    }
  });

  if (response.success) {
    state.issueStatuses.byId = {
      ...state.issueStatuses,
      ...indexById(response.payload.issueStatuses)
    };

    return {
      success: true,
      data: {
        items: response.payload.issueStatuses,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issueStatuses.length)
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.payload.limit,
      offset: response.payload.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getAll
};
