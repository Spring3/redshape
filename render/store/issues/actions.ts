import type { Context, IAction } from 'overmind';
import { Issue, PaginatedActionResponse, Response } from '../../../types';
import { indexById } from '../helpers';

type GetManyIssuesArgs = {
  filters: {
    assigned_to_id?: number | string;
    status_id?: '*' | 'open' | 'closed';
    subject?: string;
    sort?: string;
  },
  offset?: number;
  limit?: number;
}

const getMany: IAction<GetManyIssuesArgs, Promise<PaginatedActionResponse<Issue>>> = async ({ effects, state }: Context, { filters, offset = 0, limit = 20 }) => {
  const query = {
    limit: limit ? Math.abs(limit) : undefined,
    offset: offset ? Math.abs(offset) : 0,
    include: 'attachments,relations',
    ...filters
  };

  const response = await effects.mainProcess.request({
    payload: {
      route: 'issues.json',
      method: 'GET',
      query
    }
  });

  if (response.success) {
    state.issues.byId = {
      ...state.issues.byId,
      ...indexById(response.payload.issues)
    };

    console.log('payload', response.payload);

    return {
      success: true,
      data: {
        items: response.payload.issues,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issues.length),
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit,
      offset
    },
    hasMore: false,
    error: response.error
  };
};

type GetOneIssueArgs = {
  id: string;
}

const getOne: IAction<GetOneIssueArgs, Promise<Response<Issue>>> = async ({ effects, state }: Context, { id }) => {
  const response = await effects.mainProcess.request({
    payload: {
      route: `issues/${id}.json`,
      method: 'GET',
      query: {
        include: 'attachments,children,relations,journals',
      }
    }
  });

  if (response.success) {
    const issue = response.payload as Issue;
    state.issues.byId[id] = issue;
    return {
      success: true,
      data: issue
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  getOne,
  getMany
};
