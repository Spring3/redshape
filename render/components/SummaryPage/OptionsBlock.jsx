import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import actions from '../../actions';
import { Input, Label } from '../Input';

const MarginedDiv = styled.div`
  margin-top: 10px;
`;

class OptionsBlock extends Component {
  toggleClosedIssuesDisplay = () => {
    const { settingsShowClosedIssues, showClosedIssues } = this.props;
    settingsShowClosedIssues(!showClosedIssues);
  }

  toggleUseColors = () => {
    const { settingsUseColors, useColors } = this.props;
    settingsUseColors(!useColors);
  }

  render() {
    const { showClosedIssues, useColors } = this.props;
    return (
      <div>
        <Label htmlFor="queryOptions" label="Options">
          <div id="queryOptions">
            <label>
              <Input
                type="checkbox"
                checked={showClosedIssues}
                onChange={this.toggleClosedIssuesDisplay}
              />
              <span>Include Closed</span>
            </label>
          </div>
        </Label>
        <MarginedDiv>
          <label>
            <Input
              type="checkbox"
              checked={useColors}
              onChange={this.toggleUseColors}
            />
            <span>Use Colors</span>
          </label>
        </MarginedDiv>
      </div>
    );
  }
}

OptionsBlock.propTypes = {
  showClosedIssues: PropTypes.bool.isRequired,
  useColors: PropTypes.bool.isRequired,
  settingsShowClosedIssues: PropTypes.func.isRequired,
  settingsUseColors: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  showClosedIssues: state.settings.showClosedIssues,
  useColors: state.settings.useColors
});

const mapDispatchToProps = (dispatch) => ({
  settingsShowClosedIssues: (value) => dispatch(actions.settings.setShowClosedIssues(value)),
  settingsUseColors: (value) => dispatch(actions.settings.setUseColors(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(OptionsBlock);
