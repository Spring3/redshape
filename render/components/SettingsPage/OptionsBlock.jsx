import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled, { withTheme } from 'styled-components';

import Select from 'react-select';
import HelpCircleIcon from 'mdi-react/HelpCircleIcon';
import actions from '../../actions';
import { Input, Label } from '../Input';
import Tooltip from '../Tooltip';

import { availableOptions } from '../../settings';

const FlexRow = styled.div`
  display: flex;
`;

const DiscardLabel = styled.label`
  display: flex;
  align-items: center;
  margin-left: 1rem;
`;

const compSelectStyles = {
  container: base => ({ ...base, minWidth: 240 }),
};

const HelpIconStyled = styled(HelpCircleIcon)`
  padding-bottom: 1px;
`;
const LabelIcon = styled.span`
  margin-left: 0.2rem;
  color: #A0A0A0;
`;
const tooltipOptions = '- Advanced timer controls: timer and comment are modifiable at runtime.\n- Issue always editable: server will check permissions and update (or not).\n- Comments editable: needs server-side support to update notes (if valid permissions).\n- Custom Fields editable: needs server-side support to update them (if valid permissions).';
const tooltipCheckpoint = 'Save the state of the timer periodically to avoid losing temporary data\n(eg. killing/suspend/shutdown not working properly in your system).';
const OptionsInfo = (<LabelIcon><Tooltip position="right" text={tooltipOptions}><HelpIconStyled size={14} /></Tooltip></LabelIcon>);
const ProgressInfo = (<LabelIcon><Tooltip text="Use 10% unless you have server-side support."><HelpIconStyled size={14} /></Tooltip></LabelIcon>);
const TimerBehaviorInfo = (<LabelIcon><Tooltip text="Detect if the system is idle to pause the timer."><HelpIconStyled size={14} /></Tooltip></LabelIcon>);
const CheckpointInfo = (<LabelIcon><Tooltip text={tooltipCheckpoint}><HelpIconStyled size={14} /></Tooltip></LabelIcon>);

const Grid = styled.div`
  display: grid;
  grid-template-columns: auto auto 1fr;
  grid-gap: 5px;
  grid-column-gap: 80px;

  @media (max-width: 1200px) {
    grid-template-columns: auto 1fr;
  }

  @media (max-width: 950px) {
    grid-template-columns: 1fr;
  }
`;

const OptionList = styled.div`
  label {
    display: block;
  }
  label:not(:last-child) {
    margin-bottom: 0.5rem;
  }
`;

class OptionsBlock extends Component {
  constructor(props) {
    super(props);
    const {
      uiStyle, idleBehavior, progressSlider, timerCheckpoint
    } = availableOptions;
    this.options = {
      uiStyle, idleBehavior, progressSlider, timerCheckpoint
    };
  }

  toggleShowAdvancedTimerControls = () => {
    const { settingsShowAdvancedTimerControls, showAdvancedTimerControls } = this.props;
    settingsShowAdvancedTimerControls(!showAdvancedTimerControls);
  }

  toggleIdleTimeDiscard = () => {
    const { settingsIdleTimeDiscard, idleTimeDiscard } = this.props;
    settingsIdleTimeDiscard(!idleTimeDiscard);
  }

  onUiStyleChange = (uiStyle) => {
    const { settingsUiStyle, fetchAvatar, avatar_id } = this.props;
    const { value } = uiStyle;
    if (value === 'enhanced' && avatar_id >= 0) {
      fetchAvatar(avatar_id);
    }
    settingsUiStyle(value);
  }

  onIdleBehaviorChange = (idleBehavior) => {
    const { settingsIdleBehavior } = this.props;
    settingsIdleBehavior(idleBehavior.value);
  }

  onProgressSliderChange = (progressSlider) => {
    const { settingsProgressSlider } = this.props;
    settingsProgressSlider(progressSlider.value);
  }

  toggleIssueAlwaysEditable = () => {
    const { settingsIssueAlwaysEditable, isIssueAlwaysEditable } = this.props;
    settingsIssueAlwaysEditable(!isIssueAlwaysEditable);
  }

  toggleCommentsEditable = () => {
    const { settingsCommentsEditable, areCommentsEditable } = this.props;
    settingsCommentsEditable(!areCommentsEditable);
  }

  toggleCustomFieldsEditable = () => {
    const { settingsCustomFieldsEditable, areCustomFieldsEditable, getFieldsData } = this.props;
    const nowAreCustomFieldsEditable = !areCustomFieldsEditable;
    settingsCustomFieldsEditable(nowAreCustomFieldsEditable);
    if (nowAreCustomFieldsEditable) {
      getFieldsData();
    }
  }

  onTimerCheckpointChange = (timerCheckpoint) => {
    const { settingsTimerCheckpoint } = this.props;
    settingsTimerCheckpoint(timerCheckpoint.value);
  }

  render() {
    const {
      showAdvancedTimerControls, uiStyle, idleBehavior, idleTimeDiscard, progressSlider, theme, isIssueAlwaysEditable, timerCheckpoint, areCommentsEditable, areCustomFieldsEditable
    } = this.props;
    const values = {
      idleBehavior, progressSlider, uiStyle, timerCheckpoint
    };
    const { options } = this;
    const selections = Object.fromEntries(Object.entries(options).map(([key, group]) => [key, group.find(el => el.value === values[key])]));
    const idleTimeDiscardDisabled = (idleBehavior === 'none');
    return (
      <Grid>
        <Label label="Options" rightOfLabel={OptionsInfo}>
          <OptionList>
            <label>
              <Input
                type="checkbox"
                checked={showAdvancedTimerControls}
                onChange={this.toggleShowAdvancedTimerControls}
              />
              <span>Use advanced timer controls</span>
            </label>
            <label>
              <Input
                type="checkbox"
                checked={isIssueAlwaysEditable}
                onChange={this.toggleIssueAlwaysEditable}
              />
              <span>Issue is always editable</span>
            </label>
            <label>
              <Input
                type="checkbox"
                checked={areCommentsEditable}
                onChange={this.toggleCommentsEditable}
              />
              <span>Comments are editable</span>
            </label>
            <label>
              <Input
                type="checkbox"
                checked={areCustomFieldsEditable}
                onChange={this.toggleCustomFieldsEditable}
              />
              <span>Custom fields are editable</span>
            </label>
          </OptionList>
        </Label>
        <Label htmlFor="idleBehavior" label="Timer behavior" rightOfLabel={TimerBehaviorInfo}>
          <FlexRow>
            <Select
              name="idleBehavior"
              options={options.idleBehavior}
              styles={compSelectStyles}
              value={selections.idleBehavior}
              onChange={this.onIdleBehaviorChange}
              isClearable={false}
              theme={defaultTheme => ({
                ...defaultTheme,
                borderRadius: 3,
                colors: {
                  ...defaultTheme.colors,
                  primary: theme.main,
                },
              })
            }
            />
            <DiscardLabel>
              <Input
                type="checkbox"
                checked={idleTimeDiscardDisabled ? false : idleTimeDiscard}
                disabled={idleTimeDiscardDisabled}
                onChange={this.toggleIdleTimeDiscard}
              />
              <span>Auto discard idle time from timer</span>
            </DiscardLabel>
          </FlexRow>
        </Label>
        <FlexRow>
          <Label htmlFor="checkpoint" label="Timer checkpoint" rightOfLabel={CheckpointInfo}>
            <Select
              name="checkpoint"
              options={options.timerCheckpoint}
              styles={compSelectStyles}
              value={selections.timerCheckpoint}
              onChange={this.onTimerCheckpointChange}
              isClearable={false}
              theme={defaultTheme => ({
                ...defaultTheme,
                borderRadius: 3,
                colors: {
                  ...defaultTheme.colors,
                  primary: theme.main,
                },
              })
              }
            />
          </Label>
        </FlexRow>
        <FlexRow>
          <Label htmlFor="progress" label="Progress values" rightOfLabel={ProgressInfo}>
            <Select
              name="progress"
              options={options.progressSlider}
              styles={compSelectStyles}
              value={selections.progressSlider}
              onChange={this.onProgressSliderChange}
              isClearable={false}
              theme={defaultTheme => ({
                ...defaultTheme,
                borderRadius: 3,
                colors: {
                  ...defaultTheme.colors,
                  primary: theme.main,
                },
              })
              }
            />
          </Label>
        </FlexRow>
        <FlexRow>
          <Label htmlFor="uiStyle" label="UI style">
            <Select
              name="uiStyle"
              options={options.uiStyle}
              styles={compSelectStyles}
              value={selections.uiStyle}
              onChange={this.onUiStyleChange}
              isClearable={false}
              theme={defaultTheme => ({
                ...defaultTheme,
                borderRadius: 3,
                colors: {
                  ...defaultTheme.colors,
                  primary: theme.main,
                },
              })
              }
            />
          </Label>
        </FlexRow>
      </Grid>
    );
  }
}

OptionsBlock.propTypes = {
  userId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  avatar_id: PropTypes.number,
  showClosedIssues: PropTypes.bool.isRequired,
  showAdvancedTimerControls: PropTypes.bool.isRequired,
  uiStyle: PropTypes.string.isRequired,
  idleBehavior: PropTypes.string.isRequired,
  idleTimeDiscard: PropTypes.bool.isRequired,
  progressSlider: PropTypes.string.isRequired,
  isIssueAlwaysEditable: PropTypes.bool.isRequired,
  areCommentsEditable: PropTypes.bool.isRequired,
  areCustomFieldsEditable: PropTypes.bool.isRequired,
  timerCheckpoint: PropTypes.string.isRequired,
  settingsShowAdvancedTimerControls: PropTypes.func.isRequired,
  settingsUiStyle: PropTypes.func.isRequired,
  settingsIdleBehavior: PropTypes.func.isRequired,
  settingsIdleTimeDiscard: PropTypes.func.isRequired,
  settingsProgressSlider: PropTypes.func.isRequired,
  settingsIssueAlwaysEditable: PropTypes.func.isRequired,
  settingsCommentsEditable: PropTypes.func.isRequired,
  settingsCustomFieldsEditable: PropTypes.func.isRequired,
  settingsTimerCheckpoint: PropTypes.func.isRequired,
  fetchAvatar: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  userId: state.user.id,
  avatar_id: state.user.avatar_id,
  showClosedIssues: state.settings.showClosedIssues,
  showAdvancedTimerControls: state.settings.showAdvancedTimerControls,
  uiStyle: state.settings.uiStyle,
  idleBehavior: state.settings.idleBehavior,
  idleTimeDiscard: state.settings.idleTimeDiscard,
  progressSlider: state.settings.progressSlider,
  isIssueAlwaysEditable: state.settings.isIssueAlwaysEditable,
  areCommentsEditable: state.settings.areCommentsEditable,
  areCustomFieldsEditable: state.settings.areCustomFieldsEditable,
  timerCheckpoint: state.settings.timerCheckpoint,
});

const mapDispatchToProps = dispatch => ({
  settingsShowAdvancedTimerControls: value => dispatch(actions.settings.setShowAdvancedTimerControls(value)),
  settingsUiStyle: value => dispatch(actions.settings.setUiStyle(value)),
  settingsIdleBehavior: value => dispatch(actions.settings.setIdleBehavior(value)),
  settingsIdleTimeDiscard: value => dispatch(actions.settings.setIdleTimeDiscard(value)),
  settingsProgressSlider: value => dispatch(actions.settings.setProgressSlider(value)),
  settingsIssueAlwaysEditable: value => dispatch(actions.settings.setIssueAlwaysEditable(value)),
  settingsCommentsEditable: value => dispatch(actions.settings.setCommentsEditable(value)),
  settingsCustomFieldsEditable: value => dispatch(actions.settings.setCustomFieldsEditable(value)),
  settingsTimerCheckpoint: value => dispatch(actions.settings.setTimerCheckpoint(value)),
  fetchAvatar: id => dispatch(actions.user.fetchAvatar(id)),
  getFieldsData: () => dispatch(actions.fields.getAll()),
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(OptionsBlock));
