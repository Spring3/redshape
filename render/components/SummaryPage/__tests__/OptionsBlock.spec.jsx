import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';

import { SETTINGS_SHOW_CLOSED_ISSUES, SETTINGS_USE_COLORS } from '../../../actions/settings.actions';
import OptionsBlock from '../OptionsBlock';
import { getInstance, reset, initialize } from '../../../../common/request';
import theme from '../../../theme';

const redmineEndpoint = 'redmint.test.test';
const token = 'multipass';
const mockStore = configureStore([thunk]);

let axiosMock;

describe('SummaryPage => OptionsBlock component', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
    axiosMock.onAny('.json').reply(() => Promise.resolve([200]));
  });

  afterEach(() => {
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should match the snapshot', () => {
    const state = {
      settings: {
        useColors: true,
        showClosedIssues: true
      },
      user: {
        id: 1
      }
    };
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should toggle closed issues display when checkbox is clicked', () => {
    const state = {
      settings: {
        useColors: true,
        showClosedIssues: true
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    wrapper.find('input[type="checkbox"]').at(0).simulate('change', { target: { checked: false } });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_SHOW_CLOSED_ISSUES,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        showClosed: false
      }
    });
  });

  it('should toggle use of colors when checkbox is clicked', () => {
    const state = {
      settings: {
        useColors: true,
        showClosedIssues: true
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    wrapper.find('input[type="checkbox"]').at(1).simulate('change', { target: { checked: false } });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_USE_COLORS,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        useColors: false
      }
    });
  });
});
