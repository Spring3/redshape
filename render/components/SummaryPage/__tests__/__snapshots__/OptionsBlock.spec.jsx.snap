// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryPage => OptionsBlock component should match the snapshot 1`] = `
.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c5 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background: #FF7079 transition:background .2s;
  border: 2px solid transparent;
  border-radius: 3px;
  cursor: pointer;
  background: #FF7079;
  border-color: #FF7079;
}

.c5 svg {
  visibility: visible;
  position: relative;
  vertical-align: middle;
  bottom: 3px;
}

.c3:not(:disabled):hover + .c5,
.c3:focus + .c5 {
  box-shadow: 0px 0px 5px 1px #FF7F87;
}

.c3:disabled + .c5 {
  background: #F9F9F9;
  border-color: lightgrey;
}

.c2 {
  display: inline-block;
  vertical-align: middle;
}

.c2 + span {
  margin-left: 10px;
  vertical-align: middle;
}

.c0 h4 {
  margin-bottom: 10px;
  color: #A0A0A0;
}

.c1 {
  color: #A0A0A0;
}

.c6 {
  margin-top: 10px;
}

<div>
  <div
    className="form-group undefined c0"
  >
    <h4
      className="c1"
      htmlFor="queryOptions"
    >
      Options
    </h4>
    <div
      id="queryOptions"
    >
      <label>
        <div
          className="c2"
        >
          <input
            checked={true}
            className="c3 c4"
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            checked={true}
            className="c5"
          >
            <svg
              className="mdi-icon "
              fill="white"
              height="18"
              viewBox="0 0 24 24"
              width="18"
            >
              <path
                d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
              />
            </svg>
          </div>
        </div>
        <span>
          Include Closed
        </span>
      </label>
    </div>
  </div>
  <div
    className="c6"
  >
    <label>
      <div
        className="c2"
      >
        <input
          checked={true}
          className="c3 c4"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          checked={true}
          className="c5"
        >
          <svg
            className="mdi-icon "
            fill="white"
            height="18"
            viewBox="0 0 24 24"
            width="18"
          >
            <path
              d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
            />
          </svg>
        </div>
      </div>
      <span>
        Use Colors
      </span>
    </label>
  </div>
</div>
`;
